### PR TITLE
Support patching workload priority class

### PR DESCRIPTION
**Pull request by ASverdlov**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Previously, when a WorkloadPriorityClass's value was modified, existing Workloads
referencing that priority class did not have their priorities updated. This prevented
dynamic priority adjustments and potential preemptions from occurring.

This PR fixes this by introducing a WorkloadPriorityClassReconciler that watches for
priority value changes and automatically updates all affected Workloads, triggering
re-evaluation by the scheduler.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes-sigs/kueue/issues/8322.

#### Does this PR introduce a user-facing change?
```release-note
Fix a bug that WorkloadPriorityClass value changes do not trigger Workload priority updates.
```

### CODE REVIEW

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	if err := r.client.Get(ctx, req.NamespacedName, &wpc); err != nil {
+		// we'll ignore not-found errors, since there is nothing to do.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("workloadPriorityClass", klog.KObj(&wpc))
+	log.V(2).Info("Reconcile WorkloadPriorityClass")
+
+	updateCtx, cancel := context.WithTimeout(ctx, updateTimeout)
+	defer cancel()
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go]: Why timeout?

File: pkg/controller/core/workload_controller.go [hunk 2]
```diff
@@ -989,10 +990,18 @@ func (r *WorkloadReconciler) Update(e event.TypedUpdateEvent[*kueue.Workload]) b
...
 		workload.PriorityClassName(old) != "" && workload.PriorityClassName(new) != "" &&
-		workload.PriorityClassName(old) != workload.PriorityClassName(new)
+		workload.PriorityClassName(old) != workload.PriorityClassName(new) {
+		return true
+	}
+
+	// Check if priority value changed (new behavior for WorkloadPriorityClass value updates)
+	oldPriority := priority.Priority(old)
+	newPriority := priority.Priority(new)
+	return oldPriority != newPriority
```
Comment (mbobrovskyi) [inline pkg/controller/core/workload_controller.go]:
```suggestion
	return priority.Priority(old) != priority.Priority(new)
```

File: test/integration/singlecluster/controller/jobs/job/job_controller_test.go
Comment (mbobrovskyi) [inline test/integration/singlecluster/controller/jobs/job/job_controller_test.go:1]: Should we update this test?

https://github.com/kubernetes-sigs/kueue/blob/d51811804e21084954e9db5d554da2edb0e4d421/test/integration/singlecluster/controller/core/workload_controller_test.go#L476-L511

File: pkg/controller/core/workload_controller.go (hunk 2 above)
Comment (mbobrovskyi) [inline pkg/controller/core/workload_controller.go]: This could be a case for Pod Priority, right? Should we confirm that it’s the value for the Workload Priority?

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		return ctrl.Result{}, fmt.Errorf("failed to update %d/%d workloads: %v", len(updateErrors), len(workloads.Items), updateErrors[0])
+	}
+
+	log.Info("Successfully updated all workload priorities", "count", successCount)
+	return ctrl.Result{}, nil
+}
+
+func (r *WorkloadPriorityClassReconciler) Create(e event.TypedCreateEvent[*kueue.WorkloadPriorityClass]) bool {
+	log := r.log.WithValues("workloadPriorityClass", klog.KObj(e.Object))
+	log.V(2).Info("WorkloadPriorityClass create event")
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go:119]: Do we need this logs?

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	// No need to reconcile on create, as no workloads exist with this class yet
+	return false
+}
+
+func (r *WorkloadPriorityClassReconciler) Delete(e event.TypedDeleteEvent[*kueue.WorkloadPriorityClass]) bool {
+	log := r.log.WithValues("workloadPriorityClass", klog.KObj(e.Object))
+	log.V(2).Info("WorkloadPriorityClass delete event")
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go]: And here?

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+		log.V(3).Info("Priority value unchanged, skipping reconciliation")
+		return false
+	}
+
+	log.V(2).Info("Priority value changed, triggering reconciliation", "oldValue", e.ObjectOld.Value, "newValue", e.ObjectNew.Value)
+	return true
+}
+
+func (r *WorkloadPriorityClassReconciler) Generic(e event.TypedGenericEvent[*kueue.WorkloadPriorityClass]) bool {
+	r.log.V(3).Info("Got WorkloadPriorityClass generic event", "workloadPriorityClass", klog.KObj(e.Object))
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go]: Also here?

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		wl.Spec.Priority = ptr.To(wpc.Value)
+
+		if err := r.client.Update(updateCtx, wl); err != nil {
+			if apierrors.IsConflict(err) {
+				wlLog.V(2).Info("Conflict updating workload priority, will retry")
+			} else {
+				wlLog.Error(err, "Failed to update workload priority")
+			}
+			updateErrors = append(updateErrors, fmt.Errorf("workload %s/%s: %w", wl.Namespace, wl.Name, err))
+			continue
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go]: What do you think about using errors.Join() instead? After the loop, if an error exists, we can just return it.

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func (r *WorkloadPriorityClassReconciler) Create(e event.TypedCreateEvent[*kueue.WorkloadPriorityClass]) bool {
+	log := r.log.WithValues("workloadPriorityClass", klog.KObj(e.Object))
+	log.V(2).Info("WorkloadPriorityClass create event")
+	// No need to reconcile on create, as no workloads exist with this class yet
+	return false
+}
+
+func (r *WorkloadPriorityClassReconciler) Delete(e event.TypedDeleteEvent[*kueue.WorkloadPriorityClass]) bool {
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go:126]: I’m wondering what we should do on deletion. Should we remove the priority from the workload, or would it be better to add a finalizer to the WorkloadPriorityClass and remove it after all workloads are finished? This isn’t related to this task, but it’s probably something we need to consider.

/cc @olekzabl

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 6 above)
Comment (mbobrovskyi) [reply pkg/controller/core/workloadpriorityclass_controller.go]: Also, skip “not found” errors.

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 1 above)
Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go]: Removed.

File: test/integration/singlecluster/controller/jobs/job/job_controller_test.go
Comment (ASverdlov) [reply test/integration/singlecluster/controller/jobs/job/job_controller_test.go:1]: Didn't notice this one, thanks! Updated.

File: pkg/controller/core/workload_controller.go (hunk 2 above)
Comment (ASverdlov) [reply pkg/controller/core/workload_controller.go]: Done.

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 3 above)
Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go:119]: Removed.

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 4 above)
Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go]: Removed.

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 5 above)
Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go]: Removed.

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 6 above)
Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go]: I agree, we don't need this complexity. Checked how you do it in other parts of the code and updated.

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+			continue
+		}
+
+		wl.Spec.Priority = ptr.To(wpc.Value)
+
+		if err := r.client.Update(ctx, wl); err != nil {
+			if apierrors.IsConflict(err) {
+				wlLog.V(2).Info("Conflict updating workload priority, will retry")
+			} else if client.IgnoreNotFound(err) != nil {
+				wlLog.Error(err, "Failed to update workload priority")
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go:106]: I think we should skip workloads with NotFound error.

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+		if wl.Spec.Priority != nil && *wl.Spec.Priority == wpc.Value {
+			wlLog.V(3).Info("Workload priority already up to date")
+			continue
+		}
+
+		wl.Spec.Priority = ptr.To(wpc.Value)
+
+		if err := r.client.Update(ctx, wl); err != nil {
+			if apierrors.IsConflict(err) {
+				wlLog.V(2).Info("Conflict updating workload priority, will retry")
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go]: Why are we logging conflict errors but not other errors? Also why it's info?

File: pkg/controller/core/workload_controller.go [hunk 10]
```diff
@@ -989,10 +990,20 @@ func (r *WorkloadReconciler) Update(e event.TypedUpdateEvent[*kueue.Workload]) b
...
+	}
+	// Check if priority class reference changed.
+	if workload.PriorityClassName(old) != "" && workload.PriorityClassName(new) != "" &&
+		workload.PriorityClassName(old) != workload.PriorityClassName(new) {
+		return true
+	}
+	// Check if priority value changed (for WorkloadPriorityClass value updates).
+	oldPriority := priority.Priority(old)
+	newPriority := priority.Priority(new)
+	return oldPriority != newPriority
```
Comment (mbobrovskyi) [inline pkg/controller/core/workload_controller.go]:
```suggestion
	return priority.Priority(old) != priority.Priority(new)
```

These variables are used only once. To avoid allocating new variables, we can use them in place.

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+		client.MatchingFields{indexer.WorkloadPriorityClassKey: wpc.Name}); err != nil {
+		log.Error(err, "Failed to list workloads for WorkloadPriorityClass")
+		return ctrl.Result{}, err
+	}
+	if len(workloads.Items) == 0 {
+		log.V(2).Info("No workloads using this WorkloadPriorityClass")
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("Updating priority for workloads", "count", len(workloads.Items), "newValue", wpc.Value)
```
Comment (mbobrovskyi) [inline pkg/controller/core/workloadpriorityclass_controller.go]:
```suggestion
	log.V(3).Info("Updating priority for workloads", "count", len(workloads.Items), "newValue", wpc.Value)
```

Comment (mbobrovskyi) [reply pkg/controller/core/workloadpriorityclass_controller.go]: Or maybe event better remove this and add log inside loop.

log.V(3).Info("Updating priority for workload", "newValue", wpc.Value)

WDYT?

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 9 above)
Comment (mbobrovskyi) [reply pkg/controller/core/workloadpriorityclass_controller.go]: Also use for both wlLog.Error().

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 11 above)
Comment (mbobrovskyi) [reply pkg/controller/core/workloadpriorityclass_controller.go]: Ah, I see – we already have this log inside. Then I think this one should be enough.

**Review by mbobrovskyi — COMMENTED** [review]
Overall LGTM. Thank you!

**Comment by mbobrovskyi** [comment]
Could you also resolve addressed comments?

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 9 above)
Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go]: This code no longer exists, treating it generically now.

**Review by mbobrovskyi — APPROVED** [review]
Thank you!

/lgtm

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 7 above)
Comment (mimowo) [reply pkg/controller/core/workloadpriorityclass_controller.go:126]: Good question. IIUC this PR does not change the behavior in this scenario.

> Should we remove the priority from the workload

Probably not, this would break the current semantic which keeps the same priority.

>  or would it be better to add a finalizer to the WorkloadPriorityClass and remove it after all workloads are finished?

This sounds reasonable, but IIUC the only difference with the current behavior (allow to delete, but respect the old priority) is that the priority could still be mutated during termination. This seems like an edge case, so I'm ok to keep as is and wait for user feedback.

File: test/integration/singlecluster/controller/jobs/job/job_controller_test.go [hunk 12]
```diff
@@ -1547,6 +1547,224 @@ var _ = ginkgo.Describe("Interacting with scheduler", ginkgo.Ordered, ginkgo.Con
...
+			ginkgo.By("checking that the second workload is evicted", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, wlLookupKey2, createdWorkload2)).Should(gomega.Succeed())
+					g.Expect(createdWorkload2.Status.Conditions).Should(utiltesting.HaveConditionStatusTrue(kueue.WorkloadEvicted))
+					g.Expect(createdWorkload2.Status.Conditions).Should(utiltesting.HaveConditionStatusTrue(kueue.WorkloadPreempted))
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+		})
+
+		ginkgo.It("running workload priority is decreased, but still above the pending workload, and so the workload continues running undistrurbed", framework.SlowSpec, func() {
```
Comment (mimowo) [inline test/integration/singlecluster/controller/jobs/job/job_controller_test.go]:
```suggestion
		ginkgo.It("running workload priority is decreased, but still above the pending workload, and so the workload continues running undisturbed", framework.SlowSpec, func() {
```
nit

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 7 above)
Comment (mbobrovskyi) [reply pkg/controller/core/workloadpriorityclass_controller.go:126]: > This sounds reasonable, but IIUC the only difference with the current behavior (allow to delete, but respect the old priority) is that the priority could still be mutated during termination.

We could probably add a webhook to prevent changing it.

File: test/integration/singlecluster/controller/jobs/job/job_controller_test.go (hunk 12 above)
Comment (ASverdlov) [reply test/integration/singlecluster/controller/jobs/job/job_controller_test.go]: Thanks! Fixed.

File: pkg/controller/core/workloadpriorityclass_controller.go [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+			continue
+		}
+
+		wlLog.V(2).Info("Updated workload priority", "newPriority", wpc.Value)
+	}
+	return ctrl.Result{}, errors.Join(updateErrors...)
+}
+
+func (r *WorkloadPriorityClassReconciler) Create(e event.TypedCreateEvent[*kueue.WorkloadPriorityClass]) bool {
+	// No need to reconcile on create, as no workloads exist with this class yet
```
Comment (olekzabl) [inline pkg/controller/core/workloadpriorityclass_controller.go]: IIUC workloads with this WPC name can exist - if another WPC with this name existed earlier, was removed, and we're now re-creating a WPC with the same name.

Though it's a super-edge case; feel free to decide not to handle it.

**Comment by olekzabl** [comment]
Nice, thank you!

/lgtm

File: pkg/controller/core/workloadpriorityclass_controller.go (hunk 13 above)
Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go]: I stopped filtering out create events + added tests on the re-create scenario. PTAL

File: pkg/controller/core/workloadpriorityclass_controller.go
Comment (tenzen-y) [inline pkg/controller/core/workloadpriorityclass_controller.go:1]: I would request Unit Test cases for functions `Reconcile`, `Create`, `Delete`, `Update`, and `Generic` in pkg/controller/core/workloadpriorityclass_controller_test.got`.

Comment (ASverdlov) [reply pkg/controller/core/workloadpriorityclass_controller.go:1]: Done.

File: pkg/controller/core/workloadpriorityclass_controller_test.go [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta2"
+	"sigs.k8s.io/kueue/pkg/controller/core/indexer"
+	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
+	utiltestingapi "sigs.k8s.io/kueue/pkg/util/testing/v1beta2"
+)
+
+func TestWorkloadPriorityClassPredicates(t *testing.T) {
```
Comment (tenzen-y) [inline pkg/controller/core/workloadpriorityclass_controller_test.go:38]: I'd like to decouple this test into each event like `TestWorkloadPriorityClass_Create` / `TestWorkloadPriorityClass_Update` / `TestWorkloadPriorityClass_Delete` / `TestWorkloadPriorityClass_Generic` for better test tracking.

But, we can refactor this as a follow-up PR.

**Review by tenzen-y — COMMENTED** [review]
LGTM,
Thank you!
/lgtm
/approve

**Comment by tenzen-y** [comment]
/release-note-edit

```release-note
Fix a bug that WorkloadPriorityClass value changes do not trigger Workload priority updates.
```

**Comment by tenzen-y** [comment]
@ASverdlov could you manually create cherry-pick PRs into release-0.14 and release-0.15 branches?
You can use this script: https://github.com/kubernetes-sigs/kueue/blob/main/hack/cherry_pick_pull.sh

**Comment by ASverdlov** [comment]
> @ASverdlov could you manually create cherry-pick PRs into release-0.14 and release-0.15 branches? You can use this script: https://github.com/kubernetes-sigs/kueue/blob/main/hack/cherry_pick_pull.sh

Hey @tenzen-y, thanks for the reviewing the changes so swiftly!

I just tried resolving the conflicts and realised the https://github.com/kubernetes-sigs/kueue/pull/7289 PR that my changes rely on is not cherry-picked to 1.14, 1.15. So we'd need to cherry-pick those changes first. cc @mbobrovskyi

**Comment by tenzen-y** [comment]
> > @ASverdlov could you manually create cherry-pick PRs into release-0.14 and release-0.15 branches? You can use this script: https://github.com/kubernetes-sigs/kueue/blob/main/hack/cherry_pick_pull.sh
> 
> Hey @tenzen-y, thanks for the reviewing the changes so swiftly!
> 
> I just tried resolving the conflicts and realised the #7289 PR that my changes rely on is not cherry-picked to 1.14, 1.15. So we'd need to cherry-pick those changes first. cc @mbobrovskyi

Oh, I see. In that case, we can cherry-pick this PR only to release-0.15. Do you see any problem even when you cherry pick this only to 0.15?

**Comment by ASverdlov** [comment]
@tenzen-y indeed, no issues with 0.15: https://github.com/kubernetes-sigs/kueue/pull/8499

**Comment by tenzen-y** [comment]
> @tenzen-y indeed, no issues with 0.15: #8499

Awesome, thank you 👍 Let me check your PR, then.
